File: src/main/java/net/snipsniper/configwindow/UpdateButton.java
package net.snipsniper.configwindow;

import net.snipsniper.ImageManager;
import org.capturecoop.cclogger.CCLogger;
import net.snipsniper.SnipSniper;
import net.snipsniper.config.ConfigHelper;
import net.snipsniper.utils.*;
import org.capturecoop.cclogger.CCLogLevel;
import org.capturecoop.ccutils.utils.CCStringUtils;

import javax.swing.*;
import java.awt.*;

public class UpdateButton extends IDJButton {
    final String STATE_WAITING = "waiting";
    final String STATE_DOUPDATE = "update";
    final String STATE_IDLE = "idle";
    private final Image roundArrows = ImageManager.Companion.getImage("icons/roundarrows.png").getScaledInstance(16, 16, 0);
    private final Image checkmark = ImageManager.Companion.getImage("icons/checkmark.png").getScaledInstance(16, 16, 0);
    private final Image download = ImageManager.Companion.getImage("icons/download.png").getScaledInstance(16, 16, 0);

    public UpdateButton() {
        super("");
        setId(STATE_WAITING);
        setText("Check for update");
        setIcon(new ImageIcon(roundArrows));
        addActionListener(e -> {
            if(getId().equals(STATE_DOUPDATE)) {
                UpdateUtils.Companion.update();
                return;
            }

            BuildInfo bi = SnipSniper.Companion.getBuildInfo();
            Version version = bi.getVersion();
            ReleaseType updateChannel = Utils.Companion.getReleaseType(SnipSniper.Companion.getConfig().getString(ConfigHelper.MAIN.updateChannel));
            boolean isJar = SnipSniper.Companion.getPlatformType() == PlatformType.JAR;
            boolean isDev = bi.getReleaseType() == ReleaseType.DEV;
            boolean isStable = bi.getReleaseType() == ReleaseType.STABLE;

            boolean isJarAndDev = isDev && isJar;
            boolean isJarAndDevButStableBranch = updateChannel == ReleaseType.STABLE && !isStable && isJar;

            if(isJarAndDev) {
                if(getId().equals(STATE_WAITING)) {
                    setText("Checking for update...");
                    String newestHash = Utils.Companion.getShortGitHash(Utils.Companion.getHashFromAPI(Links.API_LATEST_COMMIT));
                    if(newestHash == null || newestHash.isEmpty()) {
                        setText("Error - No connection");
                        setId(STATE_WAITING);
                        setIcon(new ImageIcon(roundArrows));
                    } else if(newestHash.equals(SnipSniper.Companion.getBuildInfo().getGitHash())) {
                        setText("Up to date!");
                        setId(STATE_IDLE);
                        setIcon(new ImageIcon(checkmark));
                    } else {
                        setText(CCStringUtils.format("<html><p align='center'>Update available! (%c)</p></html>", newestHash));
                        setId(STATE_DOUPDATE);
                        setIcon(new ImageIcon(download));
                    }
                }
            } else if(isJarAndDevButStableBranch) {
                if(getId().equals(STATE_WAITING)) {
                    setText("<html><p align='center'>Switch to stable</p></html>");
                    setId(STATE_DOUPDATE);
                }
            } else {
                if(getId().equals(STATE_WAITING)) {
                    setText("Checking for update...");
                    String versionString = Utils.Companion.getTextFromWebsite(Links.STABLE_VERSION_TXT);
                    Version onlineVersion = new Version(versionString);
                    Version currentVersion = SnipSniper.Companion.getBuildInfo().getVersion();
                    if(versionString == null || versionString.isEmpty()) {
                        setText("Error - No connection");
                        setId(STATE_WAITING);
                        setIcon(new ImageIcon(roundArrows));
                    } else if (onlineVersion.equals(currentVersion) || currentVersion.isNewerThan(onlineVersion)) {
                        setText("Up to date!");
                        setId(STATE_IDLE);
                        setIcon(new ImageIcon(checkmark));
                    } else if (onlineVersion.isNewerThan(currentVersion)) {
                        if(SnipSniper.Companion.getPlatformType() == PlatformType.STEAM) {
                            setText(CCStringUtils.format("<html><p align='center'>Update available! (%c)</p><p align='center'>Check Steam to update!</p></html>", onlineVersion.digitsToString()));
                            setId(STATE_IDLE);
                        } else {
                            setText(CCStringUtils.format("<html><p align='center'>Update available! (%c)</p></html>", onlineVersion.digitsToString()));
                            setId(STATE_DOUPDATE);
                        }
                        setIcon(new ImageIcon(download));
                    } else {
                        setText("Error. Check console.");
                        CCLogger.Companion.error(String.format("Issue checking for updates. Our Version: %s, Online version: %s", currentVersion.digitsToString(), onlineVersion.digitsToString()));
                        setId(STATE_IDLE);
                        setIcon(null);
                    }
                }
            }
        });
    }
}

File: src/main/kotlin/net/snipsniper/configwindow/UpdateButton.kt
package net.snipsniper.configwindow

import net.snipsniper.SnipSniper
import net.snipsniper.config.ConfigHelper
import net.snipsniper.utils.*
import org.capturecoop.cclogger.CCLogger

class UpdateButton: IDJButton("") {
    private val STATE_WAITING = "waiting"
    private val STATE_DOUPDATE = "update"
    private val STATE_IDLE = "idle"
    private val roundArrows = "icons/roundarrows.png".getImage().scaled(16, 16)
    private val checkmark = "icons/checkmark.png".getImage().scaled(16, 16)
    private val download = "icons/download.png".getImage().scaled(16, 16)

    init {
        id = STATE_WAITING
        text = "Check for update"
        icon = roundArrows.toImageIcon()

        addActionListener {
            if (id == STATE_DOUPDATE) {
                UpdateUtils.update()
                return@addActionListener
            }
            val bi = SnipSniper.buildInfo
            val version = bi.version
            val updateChannel = Utils.getReleaseType(SnipSniper.config.getString(ConfigHelper.MAIN.updateChannel))
            val isJar = SnipSniper.platformType === PlatformType.JAR
            val isDev = bi.releaseType === ReleaseType.DEV
            val isStable = bi.releaseType === ReleaseType.STABLE
            val isJarAndDev = isDev && isJar
            val isJarAndDevButStableBranch = updateChannel === ReleaseType.STABLE && !isStable && isJar
            println(isJarAndDev)
            if (isJarAndDev) {
                if (id == STATE_WAITING) {
                    text = "Checking for update..."
                    val newestHash = Utils.getShortGitHash(Utils.getHashFromAPI(Links.API_LATEST_COMMIT)!!)
                    if (newestHash.isEmpty()) {
                        text = "Error - No connection"
                        id = STATE_WAITING
                        icon = roundArrows.toImageIcon()
                    } else if (newestHash == bi.gitHash) {
                        text = "Up to date!"
                        id = STATE_IDLE
                        icon = checkmark.toImageIcon()
                    } else {
                        text = "<html><p align='center'>Update available! ($newestHash)</p></html>"
                        id = STATE_DOUPDATE
                        icon = download.toImageIcon()
                    }
                }
            } else if (isJarAndDevButStableBranch) {
                if (id == STATE_WAITING) {
                    text = "<html><p align='center'>Switch to stable</p></html>"
                    id = STATE_DOUPDATE
                }
            } else {
                if (id == STATE_WAITING) {
                    text = "Checking for update..."
                    val versionString = Utils.getTextFromWebsite(Links.STABLE_VERSION_TXT)
                    val onlineVersion = Version(versionString!!)
                    if (versionString.isEmpty()) {
                        text = "Error - No connection"
                        id = STATE_WAITING
                        icon = roundArrows.toImageIcon()
                    } else if (onlineVersion == version || version.isNewerThan(onlineVersion)) {
                        text = "Up to date!"
                        id = STATE_IDLE
                        icon = checkmark.toImageIcon()
                    } else if (onlineVersion.isNewerThan(version)) {
                        if (SnipSniper.platformType === PlatformType.STEAM) {
                            text = "<html><p align='center'>Update available! (${onlineVersion.digitsToString()})</p><p align='center'>Check Steam to update!</p></html>"
                            id = STATE_IDLE
                        } else {
                            text = "<html><p align='center'>Update available! (${onlineVersion.digitsToString()})</p></html>"
                            id = STATE_DOUPDATE
                        }
                        icon = download.toImageIcon()
                    } else {
                        text = "Error. Check console."
                        CCLogger.error("Issue checking for updates. Our Version: ${version.digitsToString()}, Online version: ${onlineVersion.digitsToString()}")
                        id = STATE_IDLE
                        icon = null
                    }
                }
            }
        }
    }
}